### PR TITLE
Callsite override

### PR DESCRIPTION
**Pull request by Ninja3047**

### CODE REVIEW

File: lib/Passes/RemoveCallIntrinsics.cpp [hunk 1]
```diff
@@ -43,23 +38,45 @@ RemoveCallIntrinsics::runOnIntrinsic(llvm::CallInst *remillFunctionCall,
       ra.references_global_value ||  // Related to a global var/func.
       ra.references_program_counter) {  // Related to `__anvill_pc`.
 
-    // TODO(Ian): ignoring callsite decls for now
-    auto fdecl = spec.FunctionAt(ra.u.address);
+    std::optional<CallableDecl> maybe_decl;
```
Comment (2over12) [inline lib/Passes/RemoveCallIntrinsics.cpp]: Shouldnt this be initialized to spec.FunctionAt folded address ie. the default is we just call the address and dont have to rely on the override if one doesnt exist?

**Review by 2over12 — CHANGES_REQUESTED** [review]

File: lib/Passes/RemoveCallIntrinsics.cpp (hunk 1 above)
Comment (Ninja3047) [reply lib/Passes/RemoveCallIntrinsics.cpp]: https://github.com/lifting-bits/anvill/pull/398/commits/4e83da7f9ef3e614bf1b41af54ed9a6b262d1b18

**Review by 2over12 — APPROVED** [review]

File: lib/Specification.cpp [hunk 2]
```diff
@@ -394,6 +395,16 @@ Specification::DecodeFromPB(llvm::LLVMContext &context, std::istream &pb) {
   return Specification(std::move(pimpl));
 }
 
+// Return the call site at a given function address, instruction address pair, or an empty `shared_ptr`.
+std::shared_ptr<const CallSiteDecl>
+Specification::CallSiteAt(std::pair<std::uint64_t, std::uint64_t> loc) const {
```
Comment (tetsuo-cpp) [inline lib/Specification.cpp]: Don't need to copy here: `const std::pair<std::uint64_t, std::uint64_t>&`

File: lib/Passes/RemoveCallIntrinsics.cpp [hunk 3]
```diff
@@ -43,23 +38,42 @@ RemoveCallIntrinsics::runOnIntrinsic(llvm::CallInst *remillFunctionCall,
...
+        auto block_contexts = spec.GetBlockContexts();
+        if (auto bb_ctx_ref =
+                block_contexts.GetBasicBlockContextForAddr(*bb_addr);
+            bb_ctx_ref.has_value()) {
+          const auto &bb_ctx = bb_ctx_ref->get();
+          auto func = bb_ctx.GetParentFunctionAddress();
+          if (auto override_decl = spec.CallSiteAt({func, *pc_val})) {
+            DLOG(INFO) << "Overriding call site at " << std::hex << *pc_val
+                       << " in " << std::hex << func;
+            callable_decl = override_decl;
```
Comment (tetsuo-cpp) [inline lib/Passes/RemoveCallIntrinsics.cpp]: Shared pointer doesn't need to be copied: `callable_decl = std::move(override_decl)`

File: lib/Passes/RemoveCallIntrinsics.cpp [hunk 4]
```diff
@@ -43,23 +38,42 @@ RemoveCallIntrinsics::runOnIntrinsic(llvm::CallInst *remillFunctionCall,
...
+        spec.FunctionAt(ra.u.address);
+
+    if (auto pc_val =
+            GetMetadata(lifter.Options().pc_metadata_name, *remillFunctionCall);
+        pc_val.has_value()) {
+      if (auto bb_addr = GetBasicBlockAddr(f); bb_addr.has_value()) {
+        auto block_contexts = spec.GetBlockContexts();
+        if (auto bb_ctx_ref =
+                block_contexts.GetBasicBlockContextForAddr(*bb_addr);
+            bb_ctx_ref.has_value()) {
```
Comment (tetsuo-cpp) [inline lib/Passes/RemoveCallIntrinsics.cpp]: I'm not super familiar with these helpers, but is it expected that we can get a None here given that we know we're looking at a basic block function?

File: lib/Utils.cpp [hunk 5]
```diff
@@ -244,6 +244,34 @@ std::string CreateVariableName(std::uint64_t addr) {
   return ss.str();
 }
 
+std::optional<std::uint64_t> GetMetadata(llvm::StringRef tag,
+                                         llvm::Instruction &instr) {
```
Comment (tetsuo-cpp) [inline lib/Utils.cpp]: `const llvm::Instruction &`

**Review by tetsuo-cpp — APPROVED** [review]
LGTM codewise modulo some C++ nits, I'm missing context around this change though. What's this solving?

**Comment by Ninja3047** [comment]
> LGTM codewise modulo some C++ nits, I'm missing context around this change though. What's this solving?

the main use case is for fixing varargs functions
right now when we see a varargs function, we don't necessarily correctly detect the number of parameters actually used. ghidra has the ability to override a function call at a particular call site but we weren't using that information before. this pr makes it so we fetch that information (which was already in the spec) and apply it in the remove call intrinsics pass.

**Review by tetsuo-cpp — APPROVED** [review]
Awesome, thanks for the added context.

**Pull request merged**